### PR TITLE
refactor: forward ref to datepicker instance

### PR DESCRIPTION
**Pull request by ArekChr**
### Fixed Issues

$ https://github.com/Expensify/App/issues/16551
$ https://github.com/Expensify/App/issues/16551#issuecomment-1490344206

### Tests
1. Go to settings -> Profile -> Personal Details
2. Click Date of birth
3. It should open the date of birth page with opened new calendar picker

- [x] Verify that no errors appear in the JS console

### Offline tests
No offline tests needed

### QA Steps
1. Go to settings -> Profile -> Personal Details
2. Click Date of birth
3. It should open the date of birth page with opened new calendar picker

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/24796318/229082360-c21d045c-5632-4367-9ac0-0007061865e2.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/24796318/229082345-f0920cb9-a6a0-44be-9628-f96b6d7712b9.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/24796318/229082316-569d5bd9-f7d9-4eab-91a4-d0fdc160f2f0.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/24796318/229082267-eb7a158a-8c0e-427e-bcee-9c26ef9c867a.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/24796318/229082211-c7286a28-8798-4396-a0f4-76a7fa52d670.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/24796318/229082240-2404329c-0c53-47b7-8089-8f89ab03aa46.mov

</details>

### CODE REVIEW

**Comment by ArekChr** [comment]
I faced a ref forwarding issue, componentDidMount is called before the ref callback, so I wrapped it with setTimeout.
Does anyone know why this issue is happening?

**Comment by ArekChr** [comment]
@mountiny This PR is ready for review 🚀

**Review by mountiny — COMMENTED** [review]
I am not sure why this is happening, but can you try to think of other way to solve this without using setTimeout, those are usually very hacky solutions we try to avoid almost at all cost.

cc @mollfpr what do you think?

**Comment by ArekChr** [comment]
I saw some places in the code base where `InteractionManager.runAfterInteractions` is used to call a methods using ref

**Comment by ArekChr** [comment]
I tried to workaround it in many ways, but none of them helped, I'm guessing this problem may be caused by HOC

**Comment by mountiny** [comment]
Yeah `runAfterInteractions ` uses setTimeout under the hood, its discouraged to use it, but if this is the only way we are able to achieve this 🤷 

might be good to bring this to the open source channel and see if others have some ideas of why it does not work, at least we can confirm there is no other way

**Comment by ArekChr** [comment]
Good point 👌 Let's see. I asked in opensource channel

**Comment by mollfpr** [comment]
@mountiny @ArekChr I think it's because `NewDatePicker` is inside the `Form.js` component.

**Comment by mollfpr** [comment]
We can create a callback function from `Form.js` where it's called after the form is mounted, and change a state in `DateOfBirthPage` to determine if the form is ready, then we call show picker in the `componentDidUpdate` after the form is ready.

**Comment by mollfpr** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



https://user-images.githubusercontent.com/25520267/229575201-7216ed4a-4b0f-45d7-8ac8-2a5e9f6bf423.mov


</details>

<details>
<summary>Mobile Web - Chrome</summary>



https://user-images.githubusercontent.com/25520267/229575226-f9a09d75-af41-43c1-8d25-2185ec32c13e.mov


</details>

<details>
<summary>Mobile Web - Safari</summary>



https://user-images.githubusercontent.com/25520267/229575245-bcccd78b-1479-4776-a5a3-447f96abf126.mp4


</details>

<details>
<summary>Desktop</summary>



https://user-images.githubusercontent.com/25520267/229575264-58d9e65c-f254-4815-aec4-baba7e41747c.mov


</details>

<details>
<summary>iOS</summary>



https://user-images.githubusercontent.com/25520267/229575286-4c3acaf8-6461-427a-ae4f-2de6b50c8a75.mp4


</details>

<details>
<summary>Android</summary>



https://user-images.githubusercontent.com/25520267/229575758-ed401e83-2a40-438c-a85f-c2ed67f0085f.mov


</details>

**Review by mollfpr — APPROVED** [review]
LGTM 👍

All yours @mountiny!

**Review by mountiny — APPROVED** [review]
Very nice, thank you very much @mollfpr @ArekChr 

Also thank you @fedirjh for great input 🙇

File: src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js [hunk 1]
```diff
@@ -67,10 +66,7 @@ class DateOfBirthPage extends Component {
         if (params && params.year) {
             this.setState({selectedYear: params.year});
             if (this.datePicker) {
-                this.datePicker.focus();
-                if (_.isFunction(this.datePicker.click)) {
-                    this.datePicker.click();
-                }
+                this.datePicker.showPicker();
```
Comment (s77rt) [inline src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js:69]: This introduced a regression https://expensify.slack.com/archives/C049HHMV9SM/p1680625096406979 The reason is that the text input is now no longer focused so no blur event will be emitted and the calendar will not hide on clicking outside. Fix:
```diff
diff --git a/src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js b/src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js
index 4983c2cc5f..51d3dc5c55 100644
--- a/src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js
+++ b/src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js
@@ -67,6 +67,7 @@ class DateOfBirthPage extends Component {
             this.setState({selectedYear: params.year});
             if (this.datePicker) {
                 this.datePicker.showPicker();
+                this.datePicker.textInputRef.focus();
             }
         }
     }
```

Comment (mountiny) [reply src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.js:69]: Discussing in slack I dont think this is a regression, we have a different issue.

**Pull request merged**